### PR TITLE
feat: add list_and_lock_head, delete_request_lock, prolong_request_lock, batch_add_requests, batch_delete_requests and list_requests methods

### PR DESCRIPTION
**Pull request by drobnikj**
Python code after a long time, rather check it carefully. 😄

### CODE REVIEW

**Review by jirimoravcik — CHANGES_REQUESTED** [review]
I'd extend the tests, e.g. lock the requests for 10 seconds, then wait ~11 seconds and check if they were unlocked or something like that. Or maybe we'll do more extensive testing in the SDK so it's ok to have less tests here. We'll see what @fnesveda thinks.

File: tests/integration/test_request_queue.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+def random_string(length: int = 10) -> str:
+    return ''.join(random.choice(string.ascii_letters) for _ in range(length))
+
+
+def random_queue_name() -> str:
+    return f'python-client-test-queue-{random_string(5)}'
+
+
+class TestRequestQueueSync:
```
Comment (fnesveda) [inline tests/integration/test_request_queue.py:15]: Some comments and a bit more formatting would be nice here

File: src/apify_client/clients/resource_clients/request_queue.py [hunk 2]
```diff
@@ -74,6 +74,29 @@ def list_head(self, *, limit: Optional[int] = None) -> Dict:
 
         return _parse_date_fields(_pluck_data(response.json()))
 
+    def list_and_lock_head(self, *, lock_secs: int, limit: Optional[int] = None) -> Dict:
+        """Retrieve the given number of first requests from the queue and locks them for the given time.
```
Comment (fnesveda) [inline src/apify_client/clients/resource_clients/request_queue.py:78]:
```suggestion
        """Retrieve a given number of unlocked requests from the beginning of the queue and lock them for a given time.
```

File: src/apify_client/clients/resource_clients/request_queue.py [hunk 3]
```diff
@@ -74,6 +74,29 @@ def list_head(self, *, limit: Optional[int] = None) -> Dict:
...
+
+        https://docs.apify.com/api/v2#/reference/request-queues/queue-head-with-locks/get-head-and-lock
+
+        Args:
+            lock_secs (int): How long the second request will be locked for
```
Comment (fnesveda) [inline src/apify_client/clients/resource_clients/request_queue.py:83]:
```suggestion
            lock_secs (int): How long the requests will be locked for, in seconds
```

File: src/apify_client/clients/resource_clients/request_queue.py [hunk 4]
```diff
@@ -170,6 +193,50 @@ def delete_request(self, request_id: str) -> None:
...
         )
 
+    def prolong_request_lock(self, request_id: str, *, lock_secs: int, forefront: Optional[bool] = None) -> Dict:
+        """Prolong the lock on a request.
+
+        https://docs.apify.com/api/v2#/reference/request-queues/request-lock/prolong-request-lock
+
+        Args:
+            request_id (str): ID of the request to prolong the lock
+            lock_secs (int): How long to prolong the lock
```
Comment (fnesveda) [inline src/apify_client/clients/resource_clients/request_queue.py:203]:
```suggestion
            lock_secs (int): By how much to prolong the lock, in seconds
```

File: src/apify_client/clients/resource_clients/request_queue.py [hunk 5]
```diff
@@ -170,6 +193,50 @@ def delete_request(self, request_id: str) -> None:
...
+            forefront (bool, optional): Whether to put the request in the beginning or the end of the queue after lock exppires
```
Comment (fnesveda) [inline src/apify_client/clients/resource_clients/request_queue.py:204]:
```suggestion
            forefront (bool, optional): Whether to put the request in the beginning or the end of the queue after lock expires
```

File: src/apify_client/clients/resource_clients/request_queue.py [hunk 6]
```diff
@@ -170,6 +193,50 @@ def delete_request(self, request_id: str) -> None:
...
+        return _parse_date_fields(_pluck_data(response.json()))
+
+    def delete_request_lock(self, request_id: str, *, forefront: Optional[bool] = None) -> None:
+        """Delete the lock on a request.
+
+        https://docs.apify.com/api/v2#/reference/request-queues/request-lock/delete-request-lock
+
+        Args:
+            request_id (str): ID of the request to delete the lock
+            forefront (bool, optional): Whether to put the request in the beginning or the end of the queue after lock deleted
```
Comment (fnesveda) [inline src/apify_client/clients/resource_clients/request_queue.py:227]:
```suggestion
            forefront (bool, optional): Whether to put the request in the beginning or the end of the queue after the lock is deleted
```

File: src/apify_client/clients/resource_clients/request_queue.py [hunk 7]
```diff
@@ -170,6 +193,50 @@ def delete_request(self, request_id: str) -> None:
...
+
+        response = self.http_client.call(
+            url=self._url(f'requests/{request_id}/lock'),
+            method='PUT',
+            params=request_params,
+        )
+
+        return _parse_date_fields(_pluck_data(response.json()))
+
+    def delete_request_lock(self, request_id: str, *, forefront: Optional[bool] = None) -> None:
```
Comment (fnesveda) [inline src/apify_client/clients/resource_clients/request_queue.py:220]: We should have called this `unlock_request`, but too late

**Review by fnesveda — CHANGES_REQUESTED** [review]
Nice! I like that you made integration tests for this, I'm sad that we don't have them for the rest of the client 😄 I just had some notes mostly to the docstrings.

And regarding Jirka's point - I'm not sure how complicated the tests need to be here, all the aspects of the platform features should be tested by the platform integration tests, not the client tests, so I think it's fine to just check if the API endpoints work and return what they should, without too many complicated tests.

**Review by jirimoravcik — APPROVED** [review]

**Review by fnesveda — APPROVED** [review]

**Review by fnesveda — CHANGES_REQUESTED** [review]
I just realized, you should also bump the version and update the CHANGELOG.md

**Comment by jirimoravcik** [comment]
> I just realized, you should also bump the version and update the CHANGELOG.md

how about we automate this going forward?

**Comment by fnesveda** [comment]
> how about we automate this going forward?

I kind of like the versioning flexibility this gives us, plus we weren't making that many changes to these libraries for it to be worth investing the time into. But maybe there is some easy solution for this.

**Review by fnesveda — CHANGES_REQUESTED** [review]
You should bump the version to `1.3.0` in `pyproject.toml`, otherwise the publishing workflow will fail because `1.2.2` already exists.

**Pull request merged**